crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,584 @@
+        //                                 |
+        //                                 new
+        apply_and_assert_eq(
+            &mut accumulator,
+            as_vector.take(),
+            &[
+                VectorDiff::Append { values: vector!['e', 'f', 'g'] },
+                VectorDiff::Append { values: vector!['h'] },
+            ],
+        );
+
+        linked_chunk
+            .replace_gap_at(
+                ['i', 'j', 'k', 'l'],
+                linked_chunk.chunk_identifier(|chunk| chunk.is_gap()).unwrap(),
+            )
+            .unwrap();
+        assert_items_eq!(
+            linked_chunk,
+            ['a', 'w', 'x'] ['y', 'z', 'b'] ['c'] ['d'] ['i', 'j', 'k'] ['l'] ['e', 'f', 'g'] ['h']
+        );
+
+        // From an `ObservableVector` point of view, it would look like:
+        //
+        // 0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16
+        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+        // | a | w | x | y | z | b | c | d | i | j | k | l | e | f | g | h |
+        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+        //                                 ^^^^^^^^^^^^^^^^
+        //                                 |
+        //                                 new
+        apply_and_assert_eq(
+            &mut accumulator,
+            as_vector.take(),
+            &[
+                VectorDiff::Insert { index: 8, value: 'i' },
+                VectorDiff::Insert { index: 9, value: 'j' },
+                VectorDiff::Insert { index: 10, value: 'k' },
+                VectorDiff::Insert { index: 11, value: 'l' },
+            ],
+        );
+
+        linked_chunk
+            .insert_items_at(['m'], linked_chunk.item_position(|item| *item == 'a').unwrap())
+            .unwrap();
+        assert_items_eq!(
+            linked_chunk,
+            ['m', 'a', 'w'] ['x'] ['y', 'z', 'b'] ['c'] ['d'] ['i', 'j', 'k'] ['l'] ['e', 'f', 'g'] ['h']
+        );
+
+        // From an `ObservableVector` point of view, it would look like:
+        //
+        // 0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17
+        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+        // | m | a | w | x | y | z | b | c | d | i | j | k | l | e | f | g | h |
+        // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+        // ^^^^
+        // |
+        // new
+        apply_and_assert_eq(
+            &mut accumulator,
+            as_vector.take(),
+            &[VectorDiff::Insert { index: 0, value: 'm' }],
+        );
+
+        drop(linked_chunk);
+        assert!(as_vector.take().is_empty());
+
+        // Finally, ensure the “reconstitued” vector is the one expected.
+        assert_eq!(
+            accumulator,
+            vector![
+                'm', 'a', 'w', 'x', 'y', 'z', 'b', 'c', 'd', 'i', 'j', 'k', 'l', 'e', 'f', 'g', 'h'
+            ]
+        );
```

Comment:
Instead of doing this assert once at the end, could we do it in `apply_and_assert_eq`? Iterate on all the items of the linked chunk to reconstruct a `Vector` from it, and compare it against the accumulator \o/

Reply:
I think it's robust enough right now. The idea of comparing a rebuilt vector is to ensure the `VectorDiff`s we expect are the correct ones. Having this once at the end seems fine to me.